src/Common/src/System/Numerics/Hashing/Hash.cs

```diff
@@ -0,0 +1,103 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+
+namespace System.Numerics.Hashing
+{
+    internal static class Hash
+    {
+        public static int Combine(int h1, int h2)
+        {
+            // The jit optimizes this to use the ROL instruction on x86
+            // Related GitHub pull request: dotnet/coreclr#1830
+            uint shift5 = ((uint)h1 << 5) | ((uint)h1 >> 27);
+            return ((int)shift5 + h1) ^ h2;
```

Comment:
@jamesqo, where else are you expecting this is going to be used?  I can see the value in putting this one function in a shared location, but why all of the overloads that take 3-N arguments?  Those seem like they're fairly unique to a library as for tuples and would be better served just being in the tuples library as they previously were.  As you've currently factored this, any library that wants to use this function is going to need to pull in all of the rest.  Then as for this one function, I know @VSadov suggested this as an alternative to help avoid collisions... do we know that it's a net-gain performance-wise?  It may help to reduce collisions, but it would also seem to be a more complicated/expensive operation (your comment mentions it's optimized... does that make it faster than the previous hash code implementation, or just faster than the new code would have been without the optimization?)